civis_jupyter_notebooks/platform_persistence.py

```diff
@@ -45,23 +47,45 @@ def initialize_notebook_from_platform(notebook_path):
     logger.info('Notebook file ready')
 
     if hasattr(notebook_model, 'requirements_url') and notebook_model.requirements_url:
-        __pull_and_load_requirements(notebook_model.requirements_url)
+        __pull_and_load_requirements(notebook_model.requirements_url, notebook_path)
 
 
-def __pull_and_load_requirements(url):
+def __pull_and_load_requirements(url, notebook_path):
     logger.info('Pulling down the requirements file')
     r = requests.get(url)
 
     if r.status_code != 200:
         raise NotebookManagementError('Failed to pull down requirements.txt file from S3')
 
     logger.info('Writing contents of requirements file')
-    with open('requirements.txt', 'wb') as requirements:
+    requirements_path = os.path.join(os.path.dirname(notebook_path), 'requirements.txt')
+    with open(requirements_path, 'wb') as requirements:
         requirements.write(r.content)
 
     logger.info('Requirements file ready')
 
 
+def install_requirements(requirements_path):
+    while os.path.isdir(requirements_path):
+        requirements_file = os.path.join(requirements_path, 'requirements.txt')
+        logger.info('Looking for requirements at %s' % requirements_file)
+        if not os.path.isfile(requirements_file):
+            requirements_path = os.path.dirname(requirements_path)
+            continue
+
+        logger.info('Installing packages from %s' % requirements_file)
+        try:
+            subprocess.check_output(
```

Comment:
why use a subprocess? Why not just use the `pip` python package? Better error handling?

Reply:
https://media.readthedocs.org/pdf/pip/latest/pip.pdf
From section 3.15
> While it is implemented in Python, and so is available from your
Python code via import pip, you must not use pip’s internal APIs in this way.